BlueprintUI/Sources/Focus/FocusState.swift

```diff
@@ -0,0 +1,238 @@
+///                             } else if password.isEmpty {
+///                                 focusedField = .password
+///                             } else {
+///                                 handleLogin()
+///                             }
+///                         },
+///                         wrapping: Label(text: "Sign In")
+///                     )
+///                 )
+///             }
+///         }
+///     }
+///
+/// To allow for cases where focus is completely absent from a view tree, the wrapped value must be
+/// either an optional or a Boolean. Set the focus binding to `false` or `nil` as appropriate to
+/// remove focus from all bound fields. You can also use this to remove focus from a ``TextField``
+/// and thereby dismiss the keyboard.
+///
+/// ### Avoid Ambiguous Focus Bindings
+///
+/// A `TextField` can have only one focus binding, stored in its `focusBinding` property. If you apply
+/// the `focused` modifier multiple times, the last one will overwrite the previous value.
+///
+/// On the other hand, binding the same value to two views is ambiguous. In the following example,
+/// two separate fields bind focus to the `name` value:
+///
+///     struct Content: ProxyElement {
+///         enum Field: Hashable {
+///             case name
+///             case fullName
+///         }
+///
+///         @FocusState private var focusedField: Field?
+///
+///         var elementRepresentation: Element {
+///             Column { column in
+///                 column.add(
+///                     child: TextField(text: "")
+///                         .focused(when: $focusedField, equals: .name)
+///                 )
+///
+///                 column.add(
+///                     child: TextField(text: "")
+///                         .focused(when: $focusedField, equals: .name) // incorrect re-use of .name
+///                 )
+///             }
+///         }
+///     }
+///
+/// If the user moves focus to either field, the `focusedField` binding updates to `name`. However,
+/// if the app programmatically sets the value to `name`, the last field bound will be chosen.
+///
+@propertyWrapper
+public struct FocusState<Value> where Value: Hashable {
+
+    @Storage private var value: Value
+
+    /// Creates a focus state that binds to a Boolean.
+    public init() where Value == Bool {
+        value = false
+    }
+
+    /// Creates a focus state that binds to an optional type.
+    public init<T>() where Value == T?, T: Hashable {
+        value = nil
+    }
+
+    /// The current state value, taking into account whatever bindings might be
+    /// in effect due to the current location of focus.
+    ///
+    /// When focus is not in any view that is bound to this state, the wrapped
+    /// value will be `nil` (for optional-typed state) or `false` (for `Bool`-
+    /// typed state).
+    public var wrappedValue: Value {
+        get { value }
+        nonmutating set { value = newValue }
```

Comment:
I guess we don't have another option; but this property wrapper does break value semantics since it holds a reference type

Reply:
Fun trick, I'm guessing some of Apple's own wrappers do this. Seems okay for stuff like this? Is a property wrapper _really_ a struct? :P

Reply:
SwiftUI does some fancy object introspection so they can maintain value semantics; but we don't really have the ability to do it here that easily.

Reply:
> SwiftUI does some fancy object introspection so they can maintain value semantics; but we don't really have the ability to do it here that easily.

Aren't the semantics the same though? Both have a `nonmutating set`. I don't think you can say the semantics are different because of the implementation.

Reply:
Thinking about this case:

```
var myElement = MyElement() // contains an `@FocusState` internally.
var anotherElement = myElement // makes a value copy

blueprintView1.element = myElement
blueprintView2.element = anotherElement
```

These two will (AFAIK) now share the same underlying `Storage`

Reply:
Yep, I see what you mean. I'm not sure you can really do that in SwiftUI either, though.

Anchors and triggers work the same way too. I could make `FocusState` a class, but since it's typically used as a property wrapper, I don't know if that really helps.

Reply:
I wonder if we could/should set up some sort of validation for this – eg "you only get to be in one blueprint view at once" or something. Since it's insidious if you do mess it up; but maybe no one actually ever will...